Introduce waiting time track capability

ThomazFB opened the pull request:
Summary
==========
Partially fix issues #7381 and #7369 by introducing a Waiting time tracker feature that when notified with a `start` and `end` calls, will submit the elapsed time to our Tracks with the provided analytics track. This allows us to measure how long our user waits between doing any action and seeing the result. Right now, this will be used specifically for the Orders performance project, allowing us to measure the waiting time to load the Order Details and Order List views.

The work introduced here is the iOS counterpart of [this Android one](https://github.com/woocommerce/woocommerce-android/pull/7094)

How to Test
==========
1. Launch the app
2. Wait for the My Store to load
3. Go to the Orders section and wait for the Order list to load
4. Select any Order in that list and wait for the Order details to load
5. Verify in the Tracks logs that the following information is showing up:

```
🔵 Tracked dashboard_top_performers_waiting_time_loaded, properties: [AnyHashable("is_wpcom_store"): false, AnyHashable("waiting_time"): 4.4139580726623535, AnyHashable("blog_id"): ]
 ```
```
🔵 Tracked dashboard_main_stats_waiting_time_loaded, properties: [AnyHashable("blog_id"):, AnyHashable("is_wpcom_store"): false, AnyHashable("waiting_time"): 5.779787063598633]
```

```
🔵 Tracked order_detail_waiting_time_loaded, properties: [AnyHashable("is_wpcom_store"): false, AnyHashable("waiting_time"): 3.745540142059326, AnyHashable("blog_id"):]
```

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Define WaitingTimeTracker file

ThomazFB committed:
Define WaitingTimeTracker State and basic interface

ThomazFB committed:
Refactor WaitingTimeTracker state naming pattern

ThomazFB committed:
Define WaitingTimeTracker init and base properties

ThomazFB committed:
Configure onWaitingStarted in WaitingTimeTracker

ThomazFB committed:
Configure onWaitingEnded in WaitingTimeTracker

ThomazFB committed:
Fix WaitingTimeTracker enum State declaration

ThomazFB committed:
Adjust implied self reference inside the waitingTimer initialization

ThomazFB committed:
Adjust currentState type checking error

ThomazFB committed:
Configure reactive operation around WaitingTimeTracker

ThomazFB committed:
Calculate elapsedTime operation

ThomazFB committed:
Register analytics event to be triggered

ThomazFB committed:
Send analytics with elapsed time when the waiting time is finish

ThomazFB committed:
Verify if the elapsed time is between the expected range before sending it to tracks

ThomazFB committed:
Add documentation to the WaitingTimeTracker

ThomazFB committed:
Fix hound warning

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+import Foundation
+import Combine
+
+class WaitingTimeTracker {
```
Is the class `final` or is meant to be subclassed?

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+
+    /// Initialize the WaitingTimeTracker with a specific timeout, if none is provided it will set 30 seconds as the default
+    ///
+    init(waitingTimeout: TimeInterval = 30) {
```
Maybe we can extract the default timer into a constant, so it's easier to modify in the future!

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+                case .done(let waitingEndedTime):
+                    self?.sendWaitingTimeToTracks(
+                            waitingEndedTime: waitingEndedTime
+                    )
```
nit: I think this reads better in a single line. It is also a bit more aligned with the rest of the codebase 😅

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+    ///
+    /// - parameter analyticsStat: The stat to be send to Tracks when the waiting time ends
+    ///
+    func onWaitingStarted(analyticsStat: WooAnalyticsStat) {
```
Do you think just calling this `wait(stat: WooAnalyticsStat)` could read better?

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+    /// Set the Tracker state to `.done`, ending the waiting cycle. Only accepts it if the current state is `.waiting`
+    /// otherwise, it will ignore the call
+    ///
+    func onWaitingEnded() {
```
And consequently, what about `end()`?

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+            return
+        }
+
+        currentState = .done(NSDate().timeIntervalSince1970)
```
```suggestion
        currentState = .done(Date().timeIntervalSince1970)
```

We should prefer native `Swift` types, rather than legacy Objc ones!

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+
+        let elapsedTime = waitingEndedTime - waitingStartedTime
+        if 0.0...waitingTimeout ~= elapsedTime {
+            ServiceLocator.analytics.track(analyticsStat, withProperties: [
```
I think it will be better for your unit tests to inject the Analytics provider, rather than use the `ServiceLocator` one directly!

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        let elapsedTime = waitingEndedTime - waitingStartedTime
+        if 0.0...waitingTimeout ~= elapsedTime {
```
I think `0.0...waitingTimeout` will include `0` and the `timeout`. You may need to switch to a normal comparison to exclude the upper and lower bounds.

Also, I don't fully understand the waiting-timeout reason, why would we drop slow events if they take more than the timeout?

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+    /// Can be cancelled through the `waitingTimer` reference when the waiting is done.
+    ///
+    private func startWaitingTimer() {
+        waitingTimer = Timer.scheduledTimer(
+                withTimeInterval: waitingTimeout,
+                repeats: false) { [weak self] timer in
+                    self?.resetTrackerState()
+        }
```
Nit: I think having it in one line reads better in our codebase 😅 
```suggestion
        waitingTimer = Timer.scheduledTimer(withTimeInterval: waitingTimeout, repeats: false) { [weak self] timer in
            self?.resetTrackerState()
        }
```

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+        currentState = .idle
+    }
+
+    enum State: Equatable {
```
Should this enum be `private`? I don't think we want to leak this to the outside world, right?

Ecarrion commented on WooCommerce/Classes/Analytics/WaitingTimeTracker.swift:
```diff
@@ -0,0 +1,106 @@
+class WaitingTimeTracker {
+    private let waitingTimeout: TimeInterval
+
+    @Published private(set) var currentState: State = .idle
```
Do you expect this variable to be observed from the outside world? If not, I think you could use a `didSet` observation instead of using a combine observable!

Ecarrion reviewed (commented):
Hi @ThomazFB, I reviewed the code and made some suggestions! Great Job 💪  

I may be missing some of the initial requirements or reasoning but I think that we could greatly simplify this type.

- From what I see, this type is only meant to be used for one event as it does not support multiple concurrent waits. 

    If that is the case, I don't think we don't need the timeout timer, do we? We could just wait for the `done` signal and do nothing because it went over the timeout, or am I missing something?

- The iOS codebase prefers to not use `WooAnalyticsStat` directly. We use `WooAnalyticsEvent` Do you think we should update this time to support those events? which include all of the event parameters?

- It's probably too late for this, or was discarded previously, but I have to ask 😅 , could we log two events like `order_sync_started` & `order_sync_ended` and use tracks/hue/whatever to get time difference & reports?

Ecarrion commented:
Regarding unit tests, especially for types like this, I think it is better to write them before hand, so the reviewer can see to what extend it is tested and how it is meant to be used!

ThomazFB commented:
Hey @Ecarrion, thank you so much for this detailed review 🙇, before starting tackling all your suggestions, let me answer your questions since they can significantly change our approach for the code here.

> If that is the case, I don't think we don't need the timeout timer, do we? We could just wait for the done signal and do nothing because it went over the timeout, or am I missing something?

The timeout is not required. It is just a safety layer to avoid leaving the starting timestamp indefinitely there and making room for an inconsistent state or an absurd time elapsed value because of an unwanted `end` call during that view lifecycle. Do you think it's okay to just leave the timestamp registered there until the start is called again?

> The iOS codebase prefers to not use WooAnalyticsStat directly. We use WooAnalyticsEvent Do you think we should update this time to support those events? which include all of the event parameters?

Oh, I see! I will look at how to adopt the WooAnalyticsEvent for this scenario. Thanks!

> It's probably too late for this, or was discarded previously, but I have to ask 😅 , could we log two events like order_sync_started & order_sync_ended and use tracks/hue/whatever to get time difference & reports?

I thought about doing so indeed! From @shiki perspective, this would translate in significant overload in measuring everything in a query, and it would be better if we can send the elapsed time already digested. Do you think otherwise? What are your thoughts on it?

>Regarding unit tests, especially for types like this, I think it is better to write them before hand, so the reviewer can see to what extend it is tested and how it is meant to be used!

Sorry, I had to finish this one in a hurry because of our release window, but I fully agree with you and trust in a unit test culture before the review. Thanks 🙇. My main concern now is that since this type here doesn't have a clear input/output schema, I don't see how we can easily verify the internal behavior unless we leave the internal state public. Do you believe we can achieve it with a different strategy?

shiki commented:
> I thought about doing so indeed! From @shiki perspective, this would translate in significant overload in measuring everything in a query, and it would be better if we can send the elapsed time already digested. Do you think otherwise? What are your thoughts on it?

That's correct. It's difficult or maybe not reliable to do this in SQL or in our Tracks tools. The funnel for example has a minimum of 10 mins. Not great for our use case.

> Sorry, I had to finish this one in a hurry because of our release window, but I fully agree with you and trust in a unit test culture before the review.

Aside from the release window, another thing to consider is all of these might just get thrown out when we're done. Do we want to spend a lot of time on getting it tested?

Ecarrion commented:
> The timeout is not required. It is just a safety layer to avoid leaving the starting timestamp indefinitely there and making room for an inconsistent state or an absurd time elapsed value because of an unwanted `end` call during that view lifecycle. Do you think it's okay to just leave the timestamp registered there until the start is called again?

The way I see it, the type is designed as a create & dispose kind. Since we don't support concurrent waiting times, I think it will be a simpler & safer implementation if we provide an API similar like:

```swift
let orderTimeTracker = WaitingTimeTracker(event: myEvent, timeout: myTimeout)

syncEverything {
    orderTimeTracker.finish()
} 
```

That way we won't have to maintain unnecessary state within the class, nor we will need a timer.

Does that make sense? Or did I totally misunderstood the requirement?

> I thought about doing so indeed! From @shiki perspective, this would translate in significant overload in measuring everything in a query, and it would be better if we can send the elapsed time already digested. Do you think otherwise? What are your thoughts on it?

Thanks for this context, that makes sense!

 
> My main concern now is that since this type here doesn't have a clear input/output schema, I don't see how we can easily verify the internal behavior unless we leave the internal state public. Do you believe we can achieve it with a different strategy?

If you inject the analytics provider, you can easily test that your event is dispatched with the correct time_elapsed property. Feel free to reach for more concrete examples!

ThomazFB committed:
Adjust WaitingTimeTracker initialization

ThomazFB committed:
Introduce raw timestamp evaluation in WaitingTimeTracker

ThomazFB committed:
Cut out no longer needed code from WaitingTimeTracker

ThomazFB committed:
Cut out timeout parameter from the WaitingTimeTracker

ThomazFB committed:
Define currentTimeInMillis closure for better testing capabilities

ThomazFB committed:
Configure analytis service as an injected value

ThomazFB committed:
Extend WooAnalyticsEvent to support Waiting time tracks

ThomazFB committed:
Define waiting time stats in WooAnalyticsStat

ThomazFB committed:
Adjust WooAnalyticsEvent creation switch for waiting time scenarios

ThomazFB committed:
Improve WaitingTimeTracker to take advantage of WooAnalyticsEvent.WaitingTime implementation

ThomazFB committed:
Refactor WaitingTimeTracker functions

ThomazFB committed:
Update WaitingTimeTracker documentation

ThomazFB committed:
Simplify WaitingTimeTracker

ThomazFB committed:
Remove unnecessary guard let verification

ThomazFB committed:
Remove orderList scenario from WooAnalyticsEvent.WaitingTime

ThomazFB committed:
Add waiting time measurement to Order details

ThomazFB committed:
Add waiting time measurement to My Store stats and top performers

ThomazFB committed:
Create WaitingTimeTrackerTests basic structure

ThomazFB committed:
Create initial test scenarios for WaitingTimeTrackerTests

ThomazFB committed:
Complete expected WaitingTimeTrackerTests scenarios

ThomazFB committed:
Merge branch 'trunk' into issue/introduce-waiting-time-track

ThomazFB committed:
Make WaitingTimeTracker final

ThomazFB commented:
@Ecarrion Thanks again for your great and careful review. I've addressed our discussion and which ended up removing most of the change requests. The tracker is way more simplified and designed for disposal, and there's unit test coverage now. Let me know your thoughts. 

It's worth mentioning that I've also added the measurement points, so the PR is now testable by itself and contains the actual solution. I also updated the PR description to reflect that.

@shiki, just FYI and expectations syncing, I'm leaving the Order list waiting time measurement out. I noticed that we already have that data in the `orders_list_loaded` event.

```
🔵 Tracked orders_list_loaded, properties: [AnyHashable("total_duration"): 4.760702967643738, AnyHashable("is_wpcom_store"): false, AnyHashable("page_number"): 1, AnyHashable("blog_id"):, AnyHashable("status"): "", AnyHashable("date_range"): ""]
```

shiki commented:
> I'm leaving the Order list waiting time measurement out.

Are we able to easily distinguish this from subsequent page loads or pull-to-refresh?

Ecarrion commented on WooCommerce/WooCommerceTests/System/WaitingTimeTrackerTests.swift:
```diff
@@ -0,0 +1,101 @@
+        var currentTimeCallCounter = 0.0
+
+        // Given
+        let waitingTracker = WaitingTimeTracker(trackScenario: .orderDetails,
+                                                analyticsService: testAnalytics
+        ) {
+            currentTimeCallCounter += 1
+            return currentTimeCallCounter * 10
+        }
```
nit: We tend to write these in a single line, as long as it does not exceed the max linter max length (140 chars I bleieve)
```suggestion
        let waitingTracker = WaitingTimeTracker(trackScenario: .orderDetails, analyticsService: testAnalytics) {
            currentTimeCallCounter += 1
            return currentTimeCallCounter * 10
        }
```

ThomazFB replied:
Oh, thanks for the suggestion! Fixed right here: https://github.com/woocommerce/woocommerce-ios/pull/7417/commits/168c506e00ca3f8d8f1f996f27d732eace697c34

Ecarrion reviewed (approved):
Looks and works great! Good job!

ThomazFB committed:
Adjust WaitingTimeTracker creation to a single line when possible

ThomazFB commented:
> Are we able to easily distinguish this from subsequent page loads or pull-to-refresh?

@shiki Actually we won't 😬, but I believe that for us to achieve that, it would require us to do some changes in how the Order List fetches the orders data to allow us to distinguish when it's the initial first-page load from the rest. @Ecarrion may correct me in case I'm mistaken.

But do we need actually to make this differentiation? Only the network operation is getting measured here, so if there's a problem with the Orders loading, it will be reflected in everything else, like the pull to refresh and additional page loading.

shiki commented:
That's okay. It's the last day. 🙂 Time to move on.

ThomazFB committed:
Merge branch 'trunk' into issue/introduce-waiting-time-track
